<a href="https://colab.research.google.com/github/MulanSchneeberg/Deep-Learning-With-Python/blob/master/MINIST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Loading the MINIST datasets in Keras

In [30]:
from tensorflow.keras.datasets import mnist
(train_images, train_lables), (test_images, test_lables)=mnist.load_data()


The images are encoded as NumPy arrays, and the labels are an array of digits, rang
ing from 0 to 9. The images and labels have a one-to-one correspondence.

## 1.1. Take a look at the data

In [31]:
train_images.shape

(60000, 28, 28)

In [32]:
len(train_lables)

60000

In [33]:
train_lables.shape

(60000,)

In [34]:
train_lables

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [35]:
test_images.shape

(10000, 28, 28)

In [36]:
test_lables

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

# 2. The network architecture

- The core building block of neural networks is the
layer
- You can think of a layer as a filter for data: some data goes in, and it comes out in a more useful form.

- Here, our model consists of a sequence of two Dense layers, which are densely connected (also called fully connected) neural layers. The second (and last) layer is a 10-way softmax classification layer, which means it will return an array of 10 probability scores (summing to 1). Each score will be the probability that the current digit image belongs to one of our 10 digit classes.



In [37]:
from tensorflow import keras
from tensorflow.keras import layers
model= keras.Sequential([
    layers.Dense(512,activation='relu'),
    layers.Dense(10, activation='softmax')

])

# 3. Compilation: Pick Optimizer, loss function and metrices to monitor during training and test
To make the model ready for training, we need to pick three more things as part of the compilation step:

1. An optimizer
  - he mechanism through which the model will update itself based on the training data it sees, so as to improve its performance.

2. A loss function
  - How the model will be able to measure its performance on the training data, and thus how it will be able to steer itself in the right direction.

3. Metrics to monitor during training and testing
  - Here, we’ll only car



In [38]:
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# 4. Preparing the image data
Before training, we’ll preprocess the data by reshaping it into the shape the model expects and scaling it so that all values are in the [0,1] interval. Previously, our training images were stored in an array of shape (
60000,28,28) of type uint8 with values in the [0,255] interval. We’ll transform it into a float32 array of shape(60000,28*28)with values between 0 and 1.

In [39]:
train_images=train_images.reshape((60000, 28*28))

In [40]:
train_images=train_images.astype('float32')/255 # feature scaling

In [41]:
test_images=test_images.reshape((10000, 28*28))


In [42]:
test_images=test_images.astype('float32')/255

# 5. Train the model use ***fit*** method

In [43]:
model.fit(train_images, train_lables, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 3s 6ms/step - loss: 0.2634 - accuracy: 0.9233
Epoch 2/5
469/469 [==============================] - 3s 6ms/step - loss: 0.1057 - accuracy: 0.9682
Epoch 3/5
469/469 [==============================] - 3s 6ms/step - loss: 0.0700 - accuracy: 0.9793
Epoch 4/5
469/469 [==============================] - 3s 6ms/step - loss: 0.0503 - accuracy: 0.9849
Epoch 5/5
469/469 [==============================] - 3s 6ms/step - loss: 0.0379 - accuracy: 0.9886


Two quantities are displayed during training: the loss of the model over the training data, and the accuracy of the model over the training data. We quickly reach an accuracy of 0.989 (98.9%) on the training data.

# 6. Prediction using test data

In [44]:
test_digits=test_images[0:10]

In [45]:

predications=model.predict(test_digits)

In [47]:
predications[0]

array([8.7858680e-09, 4.3558880e-12, 8.0301749e-07, 5.5422239e-05,
       2.2841238e-12, 1.0048330e-08, 3.4855658e-15, 9.9994266e-01,
       6.5235170e-08, 1.0098681e-06], dtype=float32)

Each number of index i in that array corresponds to the probability that digit image test_digits[0] belongs to class i.
This first test digit has the highest probability score (0.99999106, almost 1) at index 7, so according to our model, it must be a 7

In [49]:
predications[0].argmax()

7

In [50]:
predications[0][7]

0.99994266

In [51]:
test_lables[0] # verify the finding

7

# 7. How good the model is?

In [52]:
test_loss, test_acc=model.evaluate(test_images, test_lables)

313/313 [==============================] - 1s 2ms/step - loss: 0.0652 - accuracy: 0.9801
